File: packages/guards/src/language-complexity.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface LanguageComplexityOptions {
  action: 'block' | 'warn';
  maxGradeLevel?: number;
  minGradeLevel?: number;
}

export function languageComplexity(options: LanguageComplexityOptions): Guard {
  const maxGrade = options.maxGradeLevel ?? 12;
  const minGrade = options.minGradeLevel ?? 0;

  return {
    name: 'language-complexity',
    version: '0.1.0',
    description: 'Enforces reading grade level bounds on text',
    category: 'content',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const sentences = text.split(/[.!?]+/).filter((s) => s.trim().length > 0);
      const words = text.split(/\s+/).filter(Boolean);
      const syllables = words.reduce((sum, w) => sum + Math.max(1, w.replace(/(?:[^laeiouy]es|ed|[^laeiouy]e)$/i, '').match(/[aeiouy]{1,2}/gi)?.length ?? 1), 0);
      const avgSentLen = words.length / Math.max(sentences.length, 1);
      const avgSyllables = syllables / Math.max(words.length, 1);
      const grade = Math.round(0.39 * avgSentLen + 11.8 * avgSyllables - 15.59);
      const tooHigh = grade > maxGrade;
      const tooLow = grade < minGrade;
      const triggered = tooHigh || tooLow;
      return { guardName: 'language-complexity', passed: !triggered, action: triggered ? options.action : 'allow', score: grade, latencyMs: Math.round(performance.now() - start), details: triggered ? { gradeLevel: grade, maxGrade, minGrade, reason: tooHigh ? 'too-complex' : 'too-simple' } : undefined };
    },
  };
}

File: python/open_guardrail/guards/language_complexity.py
"""Flesch-Kincaid grade level guard."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult


def _syllables(word: str) -> int:
    w = "".join(c for c in word.lower() if c.isalpha())
    if len(w) <= 3:
        return 1
    count = 0
    vowels = "aeiouy"
    prev = False
    for ch in w:
        v = ch in vowels
        if v and not prev:
            count += 1
        prev = v
    if w.endswith("e") and count > 1:
        count -= 1
    return max(1, count)


class _LanguageComplexity:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_grade_level: int = 12,
        min_grade_level: int = 0,
    ) -> None:
        self.name = "language-complexity"
        self.action = action
        self.max_grade = max_grade_level
        self.min_grade = min_grade_level

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()

        sentences = [
            s.strip()
            for s in re.split(r"[.!?]+", text)
            if s.strip()
        ]
        words = [
            w for w in text.split()
            if any(c.isalpha() for c in w)
        ]

        if not sentences or not words:
            elapsed = (time.perf_counter() - start) * 1000
            return GuardResult(
                guard_name="language-complexity",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        total_syl = sum(_syllables(w) for w in words)
        grade = (
            0.39 * (len(words) / len(sentences))
            + 11.8 * (total_syl / len(words))
            - 15.59
        )
        grade = round(grade, 1)

        triggered = (
            grade > self.max_grade
            or grade < self.min_grade
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="language-complexity",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=grade,
            message=(
                f"Grade level {grade} outside"
                f" [{self.min_grade}, {self.max_grade}]"
                if triggered else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "grade_level": grade,
                "sentences": len(sentences),
                "words": len(words),
                "syllables": total_syl,
            } if triggered else None,
        )


def language_complexity(
    *,
    action: str = "warn",
    max_grade_level: int = 12,
    min_grade_level: int = 0,
) -> _LanguageComplexity:
    return _LanguageComplexity(
        action=action,
        max_grade_level=max_grade_level,
        min_grade_level=min_grade_level,
    )
